src/web/lib/messagePicker.ts

```diff
@@ -72,15 +76,20 @@ const timeoutify = (
 	};
 };
 
-const clearAllTimeouts = (messages: CandidateConfigWithTimeout[]) =>
+const clearAllTimeouts = (messages: CandidateConfigWithTimeout<any>[]) =>
 	messages.map((m) => m.cancelTimeout());
 
 const defaultShow = () => null;
 
-type WinningMessage = {
-	idx: number;
-	meta?: any;
-};
+interface PendingMessage<T> {
+	candidateConfig: CandidateConfigWithTimeout<T>;
+	canShow: Promise<CanShowResult<T>>;
+}
+
+interface WinningMessage<T> {
+	meta: T;
+	candidate: Candidate<T>;
+}
 
 export const pickMessage = ({
```

Comment:
While fixing up `pickMessage` I did a small rewrite to pass through the `Candidate` objects with their results, rather than indexing into the array of candidates as we go.